streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreSegmentValueFormatter.java

```diff
@@ -0,0 +1,523 @@
+                    }
+                }
+
+                // prep for next iteration
+                currNextTimestamp = currTimestamp;
+                currIndex++;
+            }
+
+            throw new IllegalStateException("Search in segment expected to find result but did not.");
+        }
+
+        @Override
+        public void insertAsLatest(final long validFrom, final long validTo, final byte[] valueOrNull) {
+            final ValueAndValueSize value = new ValueAndValueSize(valueOrNull);
+
+            if (nextTimestamp > validFrom) {
+                // detected inconsistency edge case where older segment has [a,b) while newer store
+                // has [a,c), due to [b,c) having failed to write to newer store.
+                // remove entries from this store until the overlap is resolved.
+                // TODO: will be implemented in a follow-up PR
+                throw new UnsupportedOperationException("case not yet implemented");
+            }
+
+            if (nextTimestamp != validFrom) {
+                // move nextTimestamp into list as tombstone and add new record on top
+                if (isEmpty) {
+                    initializeWithRecord(new ValueAndValueSize(null), nextTimestamp, validFrom);
+                } else {
+                    insert(nextTimestamp, null, 0);
+                }
+                insertHelper(validFrom, value, 0);
+            } else {
+                // nextTimestamp is moved into segment automatically as record is added on top
+                if (isEmpty) {
+                    initializeWithRecord(value, validFrom, validTo);
+                } else {
+                    insertHelper(validFrom, value, 0);
+                }
+            }
+            // update nextTimestamp
+            nextTimestamp = validTo;
+            ByteBuffer.wrap(segmentValue, 0, TIMESTAMP_SIZE).putLong(nextTimestamp);
+        }
+
+        @Override
+        public void insertAsEarliest(final long timestamp, final byte[] valueOrNull) {
+            final ValueAndValueSize value = new ValueAndValueSize(valueOrNull);
+
+            if (isEmpty) {
+                initializeWithRecord(value, timestamp, nextTimestamp);
+            } else {
+                final int lastIndex = find(minTimestamp, false).index;
+                insertHelper(timestamp, value, lastIndex + 1);
+            }
+        }
+
+        @Override
+        public void insert(final long timestamp, final byte[] valueOrNull, final int index) {
+            final ValueAndValueSize value = new ValueAndValueSize(valueOrNull);
+            insertHelper(timestamp, value, index);
+        }
+
+        private void insertHelper(final long timestamp, final ValueAndValueSize value, final int index) {
+            if (isEmpty || index > deserIndex + 1 || index < 0) {
+                throw new IllegalArgumentException("Must invoke find() to deserialize record before insert() at specific index.");
```

Comment:
It's really just a nit and might not matter in practice. I understand both views; guess personally I would have picked `IllegalState` but both work. In the end, we hope to never hit it anyway as it would be a bug, and also, it's not user-facing.

About calling `find()` -- I am ok with not calling it -- was just wondering if it could make the usage of this class a. little simpler, but it's not a big deal. either.